server/routerlicious/packages/services-telemetry/src/lumber.ts

```diff
@@ -0,0 +1,141 @@
+    private _metadata?: ITelemetryMetadata;
+    private _durationInMs?: number;
+    private _successful?: boolean;
+    private _message?: string;
+    private _statusCode?: string;
+    private _exception?: Error;
+    private _logLevel?: LogLevel;
+    private completed = false;
+
+    public get properties(): Map<string, any> {
+        return this._properties;
+    }
+
+    public get metadata(): ITelemetryMetadata | undefined {
+        return this._metadata;
+    }
+
+    public get durationInMs(): number | undefined {
+        if (this.type === LumberType.Log) {
+            return undefined;
+        }
+        return this._durationInMs;
+    }
+
+    public get successful(): boolean | undefined {
+        if (this.type === LumberType.Log) {
+            return undefined;
+        }
+        return this._successful;
+    }
+
+    public get message(): string | undefined {
+        return this._message;
+    }
+
+    public get statusCode(): string | undefined {
+        return this._statusCode;
+    }
+
+    public get exception(): Error | undefined {
+        return this._exception;
+    }
+
+    public get logLevel(): LogLevel | undefined {
+        return this._logLevel;
+    }
+
+    constructor(
+        public readonly eventName: T,
+        public readonly type: LumberType,
+        private readonly engineList: ILumberjackEngine[],
+        properties?: Map<string, any> | Record<string, any>) {
+            if (properties) {
+                this.addProperties(properties);
+            }
+        }
+
+    public addProperty(key: string, value: any): this {
+        this._properties.set(key, value);
+        return this;
+    }
+
+    public addProperties(properties: Map<string, any> | Record<string, any>): this {
+        if (properties instanceof Map) {
+            if (this._properties.size === 0) {
+                this._properties = properties;
+            } else {
+                properties.forEach((value: any, key: string) => { this.addProperty(key, value); });
+            }
+        } else {
+            Object.entries(properties).forEach((entry) => {
+                const [key, value] = entry;
+                this.addProperty(key,value);
+            });
+        }
+        return this;
+    }
+
+    public success(
+        message: string,
+        statusCode: number | string | undefined,
+        metadata: ITelemetryMetadata,
+        logLevel: LogLevel = LogLevel.Info) {
+        this.emit(message, statusCode, metadata, logLevel, true);
+    }
+
+    public error(
+        message: string,
+        statusCode: number | string | undefined,
+        metadata: ITelemetryMetadata,
+        exception?: Error | undefined,
+        logLevel: LogLevel = LogLevel.Error) {
+        this.emit(message, statusCode, metadata, logLevel, false, exception);
+    }
+
+    private emit(
+        message: string,
+        statusCode: number | string | undefined,
+        metadata: ITelemetryMetadata,
```

Comment:
I had a chat with @TriDucNgo and we talked about the idea of having different types of "schemas" associated with each process/instance of Lumberjack. The current proposal requires a set of ITelemetryMetadata properties that might not make sense everywhere. Instead, if we could dynamically define what properties are needed in each case (so ITelemetryMetadata would look different in each scenario), then we might have a better solution - I'm currently investigating how to achieve that.

Reply:
@TriDucNgo shared some great ideas with me and I tried to implement them here. I added a new interface, `ILumberjackSchemaValidator`, plus associated implementations of the interface to this pull request to allow different schemas to be enforced by different instances of `Lumberjack`. That allows us to enforce one type of schema that requires a `tenantId`, `documentId`, `sequenceNumber`, etc, for lambdas, such as Deli and Scribe, and have a different schema for Redis, Alfred or Historian, where those properties might not be available. The schema validation is implemented through the use of classes that implement the `validate` method specified in `ILumberjackSchemaValidator`. And then, `Lumber` will use `validate` before being completed to make the properties included in the properties map are is compatible with the schema. For example, the for lambdas, the schema would be valid if a given set of properties is present, and if those properties respect some rules. That is actually an advantage of implementing the schema validator this way: we can define custom validator functions that not only check whether a property is present, but which are also able to make sure the property conforms with custom rules, such as typing, value ranges, etc.